nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/ADLSSecureCredentialsControllerService.java

```diff
@@ -0,0 +1,157 @@
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.PropertyValue;
+import org.apache.nifi.components.ValidationContext;
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.processors.azure.storage.utils.AzureStorageUtils;
+import org.apache.nifi.services.azure.keyvault.AzureKeyVaultConnectionService;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.StringJoiner;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * Provides secure credentials details for ADLS
+ *
+ * @see AbstractControllerService
+ */
+@Tags({"azure", "microsoft", "cloud", "storage", "adls", "credentials", "secure"})
+@CapabilityDescription("Defines credentials for ADLS processors.")
+public class ADLSSecureCredentialsControllerService
+        extends AbstractControllerService
+        implements ADLSCredentialsService {
+
+    private static final List<PropertyDescriptor> PROPERTIES = Collections.unmodifiableList(Arrays.asList(
+            AzureStorageUtils.KEYVAULT_CONNECTION_SERVICE,
+            AzureStorageUtils.ACCOUNT_NAME_SECRET,
+            AzureStorageUtils.ADLS_ENDPOINT_SUFFIX,
+            AzureStorageUtils.ACCOUNT_KEY_SECRET,
+            AzureStorageUtils.ACCOUNT_SAS_TOKEN_SECRET
+    ));
+
+    private ConfigurationContext context;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> results = new ArrayList<>();
+
+        boolean accountKeySet = StringUtils.isNotBlank(validationContext.getProperty(
+                AzureStorageUtils.ACCOUNT_KEY_SECRET).getValue());
+        boolean sasTokenSet = StringUtils.isNotBlank(validationContext.getProperty(
+                AzureStorageUtils.ACCOUNT_SAS_TOKEN_SECRET).getValue());
+
+        if (!onlyOneSet(accountKeySet, sasTokenSet)) {
+            StringJoiner options = new StringJoiner(", ")
+                    .add(AzureStorageUtils.ACCOUNT_KEY_SECRET.getDisplayName())
+                    .add(AzureStorageUtils.ACCOUNT_SAS_TOKEN_SECRET.getDisplayName());
+
+            results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
+                    .valid(false)
+                    .explanation("one and only one of [" + options + "] should be set")
+                    .build());
+        }
+
```

Comment:
Should we validate the non-blank secret value for non-blank keySet in your key vault service?

Reply:
maybe I misunderstood, but keyvault does not allow blank values. Let me know if you wanted me to validate something else here.